test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc

```diff
@@ -76,7 +76,7 @@ TEST_F(AwsLambdaFilterTest, DecodingHeaderStopIteration) {
  */
 TEST_F(AwsLambdaFilterTest, HeaderOnlyShouldContinue) {
   setupFilter({arn_, InvocationMode::Synchronous, true /*passthrough*/});
-  EXPECT_CALL(*signer_, sign(_));
+  EXPECT_CALL(*signer_, sign(An<Http::RequestHeaderMap&>(), false));
```

Comment:
why is this change even needed? None of the sign() calls in:

https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc

changed....

Reply:
When I updated the MockSigner, I changed the existing mock method instead of defining a new one, which caused the following compiler error from this line:

```
test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc:79:3: error: no matching member function for call to 'gmock_sign'
  EXPECT_CALL(*signer_, sign(_));
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

If I keep the old mock method, and only add a new one:
```
  MOCK_METHOD(void, sign, (Http::RequestMessage&, bool));
  MOCK_METHOD(void, sign, (Http::RequestHeaderMap&, bool));
  MOCK_METHOD(void, sign, (Http::RequestHeaderMap&));
  MOCK_METHOD(void, sign, (Http::RequestHeaderMap&, const std::string&));
```

Then the original `aws_lambda_filter_test.cc` compiles but the test fails with an unmet expectation:
```
test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc:79: Failure
Actual function call count doesn't match EXPECT_CALL(*signer_, sign(_))...
         Expected: to be called once
           Actual: never called - unsatisfied and active
```

Reply:
Ok now that I look at it, this should be enough:

```
 EXPECT_CALL(*signer_, sign(An<Http::RequestHeaderMap&>()));
```

Given that we are aiming for the call on this line:

https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc#L168

Reply:
IIUC correctly, you're suggesting this:

```diff
diff --git a/test/extensions/common/aws/mocks.h b/test/extensions/common/aws/mocks.h
index 5afc2b0bf..96962158b 100644
--- a/test/extensions/common/aws/mocks.h
+++ b/test/extensions/common/aws/mocks.h
@@ -24,6 +24,7 @@ public:
   ~MockSigner() override;
 
   MOCK_METHOD(void, sign, (Http::RequestMessage&, bool));
+  MOCK_METHOD(void, sign, (Http::RequestHeaderMap&));
   MOCK_METHOD(void, sign, (Http::RequestHeaderMap&, bool));
   MOCK_METHOD(void, sign, (Http::RequestHeaderMap&, const std::string&));
 };
diff --git a/test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc b/test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc
index 1d3cb17d1..9082f970f 100644
--- a/test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc
+++ b/test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc
@@ -76,7 +76,7 @@ TEST_F(AwsLambdaFilterTest, DecodingHeaderStopIteration) {
  */
 TEST_F(AwsLambdaFilterTest, HeaderOnlyShouldContinue) {
   setupFilter({arn_, InvocationMode::Synchronous, true /*passthrough*/});
-  EXPECT_CALL(*signer_, sign(An<Http::RequestHeaderMap&>(), false));
+  EXPECT_CALL(*signer_, sign(An<Http::RequestHeaderMap&>()));
   Http::TestRequestHeaderMapImpl input_headers;
   const auto result = filter_->decodeHeaders(input_headers, true /*end_stream*/);
   EXPECT_EQ(Http::FilterHeadersStatus::Continue, result);

```

Which results in the same failure:

```
test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc:79: Failure
Actual function call count doesn't match EXPECT_CALL(*signer_, sign(An<Http::RequestHeaderMap&>()))...
         Expected: to be called once
           Actual: never called - unsatisfied and active
```

It doesn't seem that strange to me that we have to specify the default parameters in the call we expect to make because the source call `sigv4_signer_->sign(headers)` is rewritten by the compiler as `sigv4_signer_->sign(headers, false)`. Which means that when the test is run, there really is no call to `sign` with one parameter.

Reply:
I looked back at the error you pulled from the CI logs:
```
test/extensions/filters/http/aws_lambda/aws_lambda_filter_test.cc(79,3): error: call to member function 'gmock_sign' is ambiguous
  EXPECT_CALL(*signer_, sign(An<Http::RequestHeaderMap&>(), false));
```

I remembered that I had seen this error early on, and that I resolved it using the typed matchers, so I tried replacing `false` with `Matcher<bool>(false)` and it seems to have fixed that issue. 🎉  The CI job failed again, but the failure moved from the lambda filter test to the request signing test, which has the same pattern.

I've pushed a new commit that uses Matcher<bool> for those sign calls. I don't yet know if I also need Matcher around the string arguments for the other sign calls. Since I can't reproduce locally, I'm basically throwing guesses at CI. 😕